# Day 17 - Data Wrangling

## Importing Libraries

In [1]:
# Importing Libraries
import pandas as pd 
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

## Importing titanic data

In [2]:
boat = sns.load_dataset('titanic')
print(boat.head())

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


In [3]:
boat1 = boat
boat2 = boat

In [5]:
# Simple Operations on data
#increment age by 20

(boat['age']+20).head()

0    42.0
1    58.0
2    46.0
3    55.0
4    55.0
Name: age, dtype: float64

## Find Missing Values in each column

In [6]:
boat.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [7]:
boat.shape

(891, 15)

## Dropping missing Values

In [8]:
# Removing Na values from Deck
boat.dropna(subset=['deck'], axis=0, inplace=True)
boat.shape

(203, 15)

In [9]:
boat.isnull().sum()

survived        0
pclass          0
sex             0
age            19
sibsp           0
parch           0
fare            0
embarked        2
class           0
who             0
adult_male      0
deck            0
embark_town     2
alive           0
alone           0
dtype: int64

In [10]:
# Dropping na from whole data set
boat= boat.dropna()
boat.shape

(182, 15)

In [11]:
boat.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

### Dropping rows with NA values has decreased dataset to 182 rows only. We have lost meaningful data. So instead of removing rows with NA, we will look for better ways

In [78]:
# getting boat data again
boat1 = sns.load_dataset('titanic')
print(boat1.head())

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


In [43]:
boat1.shape

(891, 15)

### Replacing Missing value with mean

In [79]:
# Finding Mean of age
mean = boat1['age'].mean()
mean

29.69911764705882

In [80]:
#Replacing NaN with Mean
boat1['age']=boat1['age'].replace(np.nan, mean )


In [81]:
boat1.isnull().sum()

survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

## Replacing NAN values of Categorical Variables
- Instead of having mean, we will replace the data with most frequent data

In [82]:
#Replacing NaN with Mode
boat1['deck'].fillna(boat1['deck'].mode()[0], inplace =True)
boat1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True    C  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False    C  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True    C  Southampton    no   True

In [83]:
#Replacing for embarked and Embarked_town

boat1['embarked'].fillna(boat1['embarked'].mode()[0], inplace =True)
boat1['embark_town'].fillna(boat1['embark_town'].mode()[0], inplace =True)

boat1.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True    C  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False    C  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True    C  Southampton    no   True

In [84]:
boat1.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

## Data Formatting

In [85]:
boat1.dtypes

survived          int64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
deck           category
embark_town      object
alive            object
alone              bool
dtype: object

In [86]:
# Changing parch in to float
boat1['parch'] = boat1['parch'].astype('float64')
boat1.dtypes

survived          int64
pclass            int64
sex              object
age             float64
sibsp             int64
parch           float64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
deck           category
embark_town      object
alive            object
alone              bool
dtype: object

In [87]:
#Changing age in to Months
boat1['age'] =boat1['age']*12
boat1['age']

0      264.000000
1      456.000000
2      312.000000
3      420.000000
4      420.000000
          ...    
886    324.000000
887    228.000000
888    356.389412
889    312.000000
890    384.000000
Name: age, Length: 891, dtype: float64

In [88]:
# Removing extra zeroes from Months
boat1['age'] = boat1['age'].astype('int64')
boat1['age']

0      264
1      456
2      312
3      420
4      420
      ... 
886    324
887    228
888    356
889    312
890    384
Name: age, Length: 891, dtype: int64

In [89]:
# Renaming Column age to Age in Months
boat1.rename(columns={'age':'age in months'}, inplace =True)
boat1

survived  pclass     sex  age in months  sibsp  parch     fare embarked  \
0           0       3    male            264      1    0.0   7.2500        S   
1           1       1  female            456      1    0.0  71.2833        C   
2           1       3  female            312      0    0.0   7.9250        S   
3           1       1  female            420      1    0.0  53.1000        S   
4           0       3    male            420      0    0.0   8.0500        S   
..        ...     ...     ...            ...    ...    ...      ...      ...   
886         0       2    male            324      0    0.0  13.0000        S   
887         1       1  female            228      0    0.0  30.0000        S   
888         0       3  female            356      1    2.0  23.4500        S   
889         1       1    male            312      0    0.0  30.0000        C   
890         0       3    male            384      0    0.0   7.7500        Q   

      class    who  adult_male deck  embark_town alive  alone  
0     Third    man        True    C  Southampton    no  False  
1     First  woman       False    C    Cherbourg   yes  False  
2     Third  woman       False    C  Southampton   yes   True  
3     First  woman       False    C  Southampton   yes  False  
4     Third    man        True    C  Southampton    no   True  
..      ...    ...         ...  ...          ...   ...    ...  
886  Second    man        True    C  Southampton    no   True  
887   First  woman       False    B  Southampton   yes   True  
888   Third  woman       False    C  Southampton    no  False  
889   First    man        True    C    Cherbourg   yes   True  
890   Third    man        True    C   Queenstown    no   True  

[891 rows x 15 columns]

## Data Normalization
- Simple Feature Scaling (dividing by Maximum value of the variable)
- Min Max Scaling ( 0-1 )
- z score scaling (-3,3) 
- Log transformation

In [90]:
# Simple Feature Scaling
boat1['fare'] = boat1['fare']/boat1['fare'].max()
boat1['fare']

0      0.014151
1      0.139136
2      0.015469
3      0.103644
4      0.015713
         ...   
886    0.025374
887    0.058556
888    0.045771
889    0.058556
890    0.015127
Name: fare, Length: 891, dtype: float64

In [93]:
# Min Max Normalization
boat1['age in months'] =(boat1['age in months']-boat1['age in months'].min()) / (boat1['age in months'].max()-boat1['age in months'].min())
boat1.head()

survived  pclass     sex  age in months  sibsp  parch      fare embarked  \
0         0       3    male       0.271204      1    0.0  0.014151        S   
1         1       1  female       0.472251      1    0.0  0.139136        C   
2         1       3  female       0.321466      0    0.0  0.015469        S   
3         1       1  female       0.434555      1    0.0  0.103644        S   
4         0       3    male       0.434555      0    0.0  0.015713        S   

   class    who  adult_male deck  embark_town alive  alone  
0  Third    man        True    C  Southampton    no  False  
1  First  woman       False    C    Cherbourg   yes  False  
2  Third  woman       False    C  Southampton   yes   True  
3  First  woman       False    C  Southampton   yes  False  
4  Third    man        True    C  Southampton    no   True

In [96]:
# z Score Normalization
boat2['age'] = (boat2['age']-boat2['age'].mean()) /  boat2['age'].std()
boat2['age'].min(), boat2['age'].max(), boat2['age']

(-2.219296474531889,
 2.8152617979568078,
 1      0.141369
 3     -0.049623
 6      1.159995
 10    -2.023211
 11     1.414651
          ...   
 871    0.714346
 872   -0.176952
 879    1.287323
 887   -1.068249
 889   -0.622600
 Name: age, Length: 203, dtype: float64)

In [97]:
# Log Normalization
boat2['fare'] = np.log(boat2['fare'])
boat2['fare'] 

/Users/asadtariq/opt/anaconda3/lib/python3.9/site-packages/pandas/core/arraylike.py:364: RuntimeWarning: divide by zero encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)


1      4.266662
3      3.972177
6      3.948596
10     2.815409
11     3.279030
         ...   
871    3.961845
872    1.609438
879    4.420746
887    3.401197
889    3.401197
Name: fare, Length: 203, dtype: float64

## Binning


In [125]:
boat3 = sns.load_dataset('titanic')
print(boat3.head())

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  


In [126]:
#Replacing NaN with Mean
boat3['age']=boat3['age'].replace(np.nan, mean )
boat3['age']

0      22.000000
1      38.000000
2      26.000000
3      35.000000
4      35.000000
         ...    
886    27.000000
887    19.000000
888    29.699118
889    26.000000
890    32.000000
Name: age, Length: 891, dtype: float64

In [127]:
# Removing extra zeroes from Months
boat3['age'] = boat3['age'].astype('int64')
boat3['age']

0      22
1      38
2      26
3      35
4      35
       ..
886    27
887    19
888    29
889    26
890    32
Name: age, Length: 891, dtype: int64

In [128]:
# Binning Age in the groups of 20 i.e 1-20 = Group1, 21-40 = Group2
bins = np.linspace(min(boat3['age']), max(boat3['age']), 4)
age_groups = ['Children', 'young', 'old']
boat3['age'] = pd.cut(boat3['age'],bins, labels=age_groups, include_lowest=True)

In [129]:
boat3['age']

0      Children
1         young
2      Children
3         young
4         young
         ...   
886       young
887    Children
888       young
889    Children
890       young
Name: age, Length: 891, dtype: category
Categories (3, object): ['Children' < 'young' < 'old']

In [132]:
# Renaming Column age to Age in Months
boat3.rename(columns={'age':'age group'}, inplace =True)
boat3.head()

survived  pclass     sex age group  sibsp  parch     fare embarked  class  \
0         0       3    male  Children      1      0   7.2500        S  Third   
1         1       1  female     young      1      0  71.2833        C  First   
2         1       3  female  Children      0      0   7.9250        S  Third   
3         1       1  female     young      1      0  53.1000        S  First   
4         0       3    male     young      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

## Converting Categorical in to Numeric data

In [109]:
#Changing Gender in to 0-1
pd.get_dummies(boat3, columns = ['sex'])


survived  pclass   age  sibsp  parch     fare embarked   class    who  \
0           0       3  22.0      1      0   7.2500        S   Third    man   
1           1       1  38.0      1      0  71.2833        C   First  woman   
2           1       3  26.0      0      0   7.9250        S   Third  woman   
3           1       1  35.0      1      0  53.1000        S   First  woman   
4           0       3  35.0      0      0   8.0500        S   Third    man   
..        ...     ...   ...    ...    ...      ...      ...     ...    ...   
886         0       2  27.0      0      0  13.0000        S  Second    man   
887         1       1  19.0      0      0  30.0000        S   First  woman   
888         0       3   NaN      1      2  23.4500        S   Third  woman   
889         1       1  26.0      0      0  30.0000        C   First    man   
890         0       3  32.0      0      0   7.7500        Q   Third    man   

     adult_male deck  embark_town alive  alone  sex_female  sex_male  
0          True  NaN  Southampton    no  False           0         1  
1         False    C    Cherbourg   yes  False           1         0  
2         False  NaN  Southampton   yes   True           1         0  
3         False    C  Southampton   yes  False           1         0  
4          True  NaN  Southampton    no   True           0         1  
..          ...  ...          ...   ...    ...         ...       ...  
886        True  NaN  Southampton    no   True           0         1  
887       False    B  Southampton   yes   True           1         0  
888       False  NaN  Southampton    no  False           1         0  
889        True    C    Cherbourg   yes   True           0         1  
890        True  NaN   Queenstown    no   True           0         1  

[891 rows x 16 columns]